AsyncDisplayKit/ASCollectionView.mm

```diff
@@ -340,6 +350,43 @@ - (void)collectionView:(UICollectionView *)collectionView didEndDisplayingCell:(
 }
 
 
+#pragma mark -
+#pragma mark Batch Fetching
+
+- (void)scrollViewWillEndDragging:(UIScrollView *)scrollView withVelocity:(CGPoint)velocity targetContentOffset:(inout CGPoint *)targetContentOffset
+{
+  [self handleBatchFetchScrollingToOffset:*targetContentOffset];
+
+  if ([_asyncDelegate respondsToSelector:@selector(scrollViewWillEndDragging:withVelocity:targetContentOffset:)]) {
+    [_asyncDelegate scrollViewWillEndDragging:scrollView withVelocity:velocity targetContentOffset:targetContentOffset];
+  }
+}
+
+- (BOOL)shouldFetchBatch
+{
+  if ([self.asyncDelegate respondsToSelector:@selector(shouldBatchFetchForCollectionView:)]) {
+    return [self.asyncDelegate shouldBatchFetchForCollectionView:self];
+  } else {
+    // if the delegate does not respond to this method, there is no point in starting to fetch
+    return [self.asyncDelegate respondsToSelector:@selector(collectionView:beginBatchFetchingWithContext:)];
+  }
+}
+
+- (void)handleBatchFetchScrollingToOffset:(CGPoint)targetOffset
+{
+  ASDisplayNodeAssert(_batchContext != nil, @"Batch context should exist");
+
+  if (![self shouldFetchBatch]) {
+    return;
+  }
+
+  if (ASDisplayShouldFetchBatchForContext(_batchContext, [self scrollDirection], self.bounds, self.contentSize, targetOffset, _leadingScreensForBatching)) {
+    [_batchContext beginBatchFetching];
+    [self.asyncDelegate collectionView:self beginBatchFetchingWithContext:_batchContext];
```

Comment:
we should either dispatch_async() this to a general concurrent background queue, or warn folks in the header that this method is called on main and that they should trampoline elsewhere